routing/index_graph_starter.cpp

```diff
@@ -1,180 +1,396 @@
   // start fake, finish fake and at least one normal nearest segment.
   CHECK_GREATER_OR_EQUAL(segments.size(), 3, ());
-  CHECK_EQUAL(segments.front(), kStartFakeSegment, ());
-  CHECK_EQUAL(segments.back(), kFinishFakeSegment, ());
+  CHECK(IsFakeSegment(segments.front()), ());
+  CHECK(IsFakeSegment(segments.back()), ());
 }
 
-// static
-vector<Segment>::const_iterator IndexGraphStarter::GetNonFakeStartIt(vector<Segment> const & segments)
+set<NumMwmId> IndexGraphStarter::GetMwms() const
 {
-  CheckValidRoute(segments);
-  // See CheckValidRoute comment.
-  return segments.begin() + 1;
-}
+  set<NumMwmId> mwms;
+  for (auto const & kv : m_fake.m_fakeToReal)
+    mwms.insert(kv.second.GetMwmId());
 
-// static
-vector<Segment>::const_iterator IndexGraphStarter::GetNonFakeFinishIt(vector<Segment> const & segments)
-{
-  CheckValidRoute(segments);
-  // See CheckValidRoute comment.
-  return segments.end() - 2;
+  return mwms;
 }
 
 // static
 size_t IndexGraphStarter::GetRouteNumPoints(vector<Segment> const & segments)
 {
   CheckValidRoute(segments);
-  return segments.size() - 1;
+  return segments.size() + 1;
 }
 
 Junction const & IndexGraphStarter::GetRouteJunction(vector<Segment> const & segments,
-                                                     size_t pointIndex)
+                                                     size_t pointIndex) const
 {
-  if (pointIndex == 0)
-    return m_start.GetJunction();
+  CHECK(!segments.empty(), ());
+  CHECK_LESS_OR_EQUAL(
+      pointIndex, segments.size(),
+      ("Point with index", pointIndex, "does not exist in route with size", segments.size()));
+  if (pointIndex == segments.size())
+    return GetJunction(segments[pointIndex - 1], true);
+  return GetJunction(segments[pointIndex], false);
+}
+
+double IndexGraphStarter::CalcSegmentWeight(Segment const & segment) const
+{
+  if (!IsFakeSegment(segment))
+  {
+    return m_graph.GetEstimator().CalcSegmentWeight(
+        segment, m_graph.GetRoadGeometry(segment.GetMwmId(), segment.GetFeatureId()));
+  }
 
-  if (pointIndex + 1 == GetRouteNumPoints(segments))
-    return m_finish.GetJunction();
+  auto const fakeVertexIt = m_fake.m_segmentToVertex.find(segment);
 
-  CHECK_LESS(pointIndex, segments.size(), ());
-  return GetJunction(segments[pointIndex], true /* front */);
+  CHECK(fakeVertexIt != m_fake.m_segmentToVertex.end(),
+        ("Requested junction for invalid fake segment."));
+  return m_graph.GetEstimator().CalcLeapWeight(fakeVertexIt->second.GetPointFrom(),
+                                               fakeVertexIt->second.GetPointTo());
 }
 
-void IndexGraphStarter::GetEdgesList(Segment const & segment, bool isOutgoing,
-                                     vector<SegmentEdge> & edges)
+double IndexGraphStarter::CalcRouteSegmentWeight(vector<Segment> const & route,
+                                                 size_t segmentIndex) const
 {
-  edges.clear();
+  CHECK(!route.empty(), ());
```

Comment:
This check is redundant in presence of the second check.

Reply:
removed